# Imports

In [1]:
import pandas as pd
import numpy as np

pd.set_option('max_rows', 200)
pd.set_option('max_colwidth', 400)

# Loading of data

In [2]:
idigbio_dataset_root_path = "/content/drive/MyDrive/Tesis/Datasets/idigbio"

In [6]:
idigbio_ocurrence = pd.read_csv(f"{idigbio_dataset_root_path}/occurrence.csv",
                                sep=",")

# We don't need the idigbio:flags column, so let's drop it
idigbio_ocurrence.drop(["idigbio:flags"],
                       axis=1,
                       inplace=True)

display(idigbio_ocurrence.head(5))

print(idigbio_ocurrence.shape)

coreid  ...  dwc:waterBody
0  9dfd5d94-7f54-4e9f-880e-e683eae3a91c  ...            NaN
1  65c3037a-ea4b-4baa-81eb-8981c9a18b62  ...            NaN
2  19f0db9c-e284-441a-b6ed-8e5a693ba506  ...            NaN
3  0d53c48d-e5ad-43ba-8465-271b34cbdda9  ...            NaN
4  a64aca3f-d22d-497e-9f18-74d2a2ade357  ...            NaN

[5 rows x 80 columns]

(6739, 80)


In [8]:
idigbio_multimedia = pd.read_csv(f"{idigbio_dataset_root_path}/multimedia.csv",
                                 sep=",")

display(idigbio_multimedia.head(5))

print(idigbio_multimedia.shape)

coreid  ... exif:PixelYDimension
0  1101d707-b36b-4258-98b8-b32a898ac593  ...                  NaN
1  e3b87182-567f-499d-a734-587675e005cc  ...                  NaN
2  8ac1c18a-bf3c-453e-b398-e3c360db8aba  ...               6739.0
3  8f4153e1-51fa-40f1-b9a2-326816aae16f  ...               6743.0
4  1a439d0f-4092-4088-8c86-8a4c9dc0bf71  ...               6738.0

[5 rows x 22 columns]

(7190, 22)


# EDA

Let's check whether all of the records have been identified.

### Mexican species

In [9]:
mexican_species = ['Solanum Acerifolium',
 'Solanum Aculeatissimum',
 'Solanum Agrimonifolium',
 'Solanum Aligerum',
 'Solanum Americanum',
 'Solanum Angustifolium',
 'Solanum Aphyodendron',
 'Solanum Appendiculatum',
 'Solanum Atitlanum',
 'Solanum Aturense',
 'Solanum Aviculare',
 'Solanum Axillifolium',
 'Solanum Betaceum',
 'Solanum Bicorne',
 'Solanum Brevipedicellatum',
 'Solanum Bulbocastanum',
 'Solanum Campechiense',
 'Solanum Candidum',
 'Solanum Capsicoides',
 'Solanum Cardiophyllum',
 'Solanum Caripense',
 'Solanum Chiapasense',
 'Solanum Chrysotrichum',
 'Solanum Citrinum',
 'Solanum Citrullifolium',
 'Solanum Clarum',
 'Solanum Cobanense',
 'Solanum Cordovense',
 'Solanum Corymbosum',
 'Solanum Dasyanthum',
 'Solanum Dasyneuron',
 'Solanum Davisense',
 'Solanum Deflexum',
 'Solanum Demissum',
 'Solanum Dimidiatum',
 'Solanum Diphyllum',
 'Solanum Donianum',
 'Solanum Dulcamaroides',
 'Solanum Edinense',
 'Solanum Edmundoi',
 'Solanum Ehrenbergii',
 'Solanum Elaeagnifolium',
 'Solanum Erianthum',
 'Solanum Erythrotrichum',
 'Solanum Ferrugineum',
 'Solanum Fructo-Tecto',
 'Solanum Gardneri',
 'Solanum Glaucescens',
 'Solanum Grayi',
 'Solanum Guerreroense',
 'Solanum Hazenii',
 'Solanum Heterodoxum',
 'Solanum Hindsianum',
 'Solanum Hintonii',
 'Solanum Hirtum',
 'Solanum Hjertingii',
 'Solanum Hougasii',
 'Solanum Houstonii',
 'Solanum Ionidium',
 'Solanum Iopetalum',
 'Solanum Jamaicense',
 'Solanum Jamesii',
 'Solanum Johnstonii',
 'Solanum Knoblochii',
 'Solanum Lanceifolium',
 'Solanum Lanceolatum',
 'Solanum Laxum',
 'Solanum Lepidotum',
 'Solanum Lesteri',
 'Solanum Leucandrum',
 'Solanum Lignescens',
 'Solanum Lumholtzianum',
 'Solanum Lycopersicum',
 'Solanum Malacothrix',
 'Solanum Mammosum',
 'Solanum Marginatum',
 'Solanum Michoacanum',
 'Solanum Mitlense',
 'Solanum Morelliforme',
 'Solanum Muenscheri',
 'Solanum Muricatum',
 'Solanum Myriacanthum',
 'Solanum Narcoticosmum',
 'Solanum Nigrescens',
 'Solanum Nigricans',
 'Solanum Nitidibaccatum',
 'Solanum Nudum',
 'Solanum Oxycarpum',
 'Solanum Palmillae',
 'Solanum Pectinatum',
 'Solanum Phaseoloides',
 'Solanum Pinnatisectum',
 'Solanum Plumense',
 'Solanum Polyadenium',
 'Solanum Pruinosum',
 'Solanum Pseudocapsicum',
 'Solanum Pubigerum',
 'Solanum Pulverulentifolium',
 'Solanum Refractum',
 'Solanum Rojasianum',
 'Solanum Rostratum',
 'Solanum Rovirosanum',
 'Solanum Rudepannum',
 'Solanum Sambucinum',
 'Solanum Schenckii',
 'Solanum Schlechtendalianum',
 'Solanum Seaforthianum',
 'Solanum Setigeroides',
 'Solanum Sisymbriifolium',
 'Solanum Skutchii',
 'Solanum Sousae',
 'Solanum Splendens',
 'Solanum Stenophyllidium',
 'Solanum Stoloniferum',
 'Solanum Suaveolens',
 'Solanum Subvelutinum',
 'Solanum Tacanense',
 'Solanum Tampicense',
 'Solanum Tarnii',
 'Solanum Tenuipes',
 'Solanum Torvum',
 'Solanum Tribulosum',
 'Solanum Trifidum',
 'Solanum Triquetrum',
 'Solanum Triunfense',
 'Solanum Trizygum',
 'Solanum Tuerckheimii',
 'Solanum Umbellatum',
 'Solanum Umbelliferum',
 'Solanum Vallis-Mexici',
 'Solanum Verrucosum',
 'Solanum Volubile',
 'Solanum Wendlandii',
 'Solanum Wrightii']

 mexican_species = [sp.split()[1].lower() for sp in mexican_species]

## Occurrences

In [10]:
idigbio_ocurrence[["coreid", "dwc:specificEpithet"]].isna().sum()

coreid                   0
dwc:specificEpithet    220
dtype: int64

In [27]:
idigbio_ocurrence[(idigbio_ocurrence["dwc:specificEpithet"].isna()) | (idigbio_ocurrence["dwc:specificEpithet"].str.lower() == "sp.")].shape

(241, 80)

We'll hold off these 241 records for the test dataset. For now, let's figure out how many records have been identified.



In [23]:
idigbio_ocurrence_identified = idigbio_ocurrence[(idigbio_ocurrence["dwc:specificEpithet"].isna() == False) & (idigbio_ocurrence["dwc:specificEpithet"].str.lower() != "sp.")]
print(idigbio_ocurrence_identified.shape)

(6498, 80)


Next, let's see how many unique species we have in the dataset, let's remember that as of 2021, there are 134 identified species in the Solanum Genus for Mexico.

In [24]:
unique_species = idigbio_ocurrence_identified["dwc:specificEpithet"].unique()
print("\n".join(sorted(unique_species)))
print(unique_species.shape)

accrescens
acerifolium
aculeatissimum
aculeolatum
adscendens
agrarium
agrimoniifolium
aligerum
americanum
amictum
andrieuxi
angustifolium
aphyodendron
appendiculatum
asperolanatum
aturense
axillifolium
bahamense
bejarense
benthami
betaceum
bombense
bonariense
brevipedicellatum
bulbocastanum
calvum
campechiense
candidum
capsicoides
cardiophyllum
chacoense
chenopodioides
chiapasense
chiapensis
chrysotrichum
circinatum
citrinum
citrullifolia
clarum
cobanense
colombianum
confusum
conioides
cordifolium
cordovense
cornutum
coronopus
corymbosum
dasyadenium
dasyanthum
dasyneuron
deflexum
dejectum
delilei
demissum
dichotomum
dimidiatum
diphyllum
diversifolium
donianum
donnell-smithii
douglasii
dulcamara
dulcamaroides
dumosum
edinense
ehrenbergii
elaeagnifolium
ensifolium
erianthum
erythrotrichum
extusviolascens
ferrugineum
fraxinifolium
fructo-tecto
fructu-tecto
furcatum
gardneri
glaucescens
glutinosum
grayi
hazenii
heterodoxum
hindsianum
hintonii
hirtum
hjertingii
hougasii
houstonii
indet
inte

There are species named *fructu-tecto* which are a typo, they should be labeled as *fructo-tecto*, let's correct that. *Agrimonifolium* and *Citrullifolium* are other examples of typos in the species name, so let's correct that.

In [67]:
idigbio_ocurrence_identified.loc[idigbio_ocurrence_identified["dwc:specificEpithet"].str.lower() == "fructu-tecto", 'dwc:specificEpithet'] = "fructo-tecto"
idigbio_ocurrence_identified.loc[idigbio_ocurrence_identified["dwc:specificEpithet"].str.lower() == "agrimoniifolium", 'dwc:specificEpithet'] = "agrimonifolium"
idigbio_ocurrence_identified.loc[idigbio_ocurrence_identified["dwc:specificEpithet"].str.lower() == "citrullifolia", 'dwc:specificEpithet'] = "citrullifolium"


# Refresh the list of unique species
unique_species = idigbio_ocurrence_identified["dwc:specificEpithet"].unique()

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:1763: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)


### Missing species

We have 202 species even though only 134 species have been identified for Mexico, this is happening because:

1. Some names for the species are invalid/outdated/have been replaced
1. Species are outright misidentified

Let's first figure out if we have at least one record of the 134 valid species for Mexico. 

In [74]:
missing_species = [sp for sp in mexican_species if sp not in unique_species]
print("\n".join(missing_species))
print(f"Missing species: {len(missing_species)}")

atitlanum
aviculare
bicorne
caripense
davisense
edmundoi
guerreroense
knoblochii
nitidibaccatum
setigeroides
splendens
triunfense
Missing species: 12


### Invalid species

Let's get all of the invalid species in the dataset.

In [75]:
invalid_species = sorted([sp for sp in idigbio_ocurrence_identified["dwc:specificEpithet"].unique() if sp.lower() not in mexican_species])

print("\n".join(invalid_species))
print(len(invalid_species))

accrescens
adscendens
agrarium
andrieuxi
asperolanatum
bahamense
bejarense
benthami
bombense
bonariense
calvum
chacoense
chenopodioides
chiapensis
colombianum
confusum
conioides
cordifolium
cornutum
coronopus
dasyadenium
dejectum
delilei
dichotomum
diversifolium
donnell-smithii
douglasii
dulcamara
dumosum
ensifolium
extusviolascens
fraxinifolium
furcatum
glutinosum
indet
interius
lambii
latiflorum
lentum
leptopodum
lichtensteinii
longifolium
macrotomum
mexiae
molinum
morellifolium
nigrum
nocturnum
nutans
ochraceo-ferrugineum
ovandense
pedunculare
pilosiusculum
pimpinellifolium
pringlei
ptychanthum
purpusii
rude-pannum
salsum
salviaefolium
sarrachoides
schlechtenthalianum
sieberi
sinuatum
slodgetti
subinerme
tricuspidatum
triste
tuberosum
ulmoides
uncinellum
undesignated
uniflorum
wallacei
74


Like we said, these species have either been renamed or they don't really belong to Mexico, so let's create a mapping of actual names to replace these with.

For those species names that don't belong to Mexico, we'll remove the records from the dataset.

In [71]:
invalid_mappings = {
    "aculeolatum": "refractum",
    "amictum": "lanceolatum",
    "blodgettii": "donianum",
    "circinatum": "splendens",
    "hernandesii": "ferrugineum",
    "palmeri": "umbelliferum",
    "tenuilobatum": "umbelliferum",
    "tridynamum": "houstonii",
    "xanti": "umbelliferum"
}

In [73]:
# Correct incorrect entries
idigbio_ocurrence_identified["dwc:specificEpithet"] = idigbio_ocurrence_identified["dwc:specificEpithet"].replace(invalid_mappings)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [76]:
idigbio_ocurrence_identified["dwc:specificEpithet"] = idigbio_ocurrence_identified["dwc:specificEpithet"].str.lower()

non_mexican_idx = idigbio_ocurrence_identified[~idigbio_ocurrence_identified["dwc:specificEpithet"].isin(mexican_species)].index

idigbio_ocurrence_identified.drop(non_mexican_idx, inplace=True)

idigbio_ocurrence_identified.head(5)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4174: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


coreid  ...  dwc:waterBody
0  9dfd5d94-7f54-4e9f-880e-e683eae3a91c  ...            NaN
1  65c3037a-ea4b-4baa-81eb-8981c9a18b62  ...            NaN
2  19f0db9c-e284-441a-b6ed-8e5a693ba506  ...            NaN
3  0d53c48d-e5ad-43ba-8465-271b34cbdda9  ...            NaN
4  a64aca3f-d22d-497e-9f18-74d2a2ade357  ...            NaN

[5 rows x 80 columns]

In [77]:
print(idigbio_ocurrence_identified.shape)
idigbio_ocurrence_identified["dwc:specificEpithet"].value_counts()

(6061, 80)


lanceolatum           365
erianthum             240
pubigerum             239
americanum            239
stoloniferum          238
nigrescens            202
ferrugineum           190
nigricans             187
houstonii             177
elaeagnifolium        169
nudum                 164
umbellatum            154
rostratum             153
hindsianum            148
chrysotrichum         124
torvum                122
demissum              116
diphyllum             113
rudepannum            104
aphyodendron          104
verrucosum             98
cardiophyllum          92
grayi                  80
refractum              78
schlechtendalianum     77
dulcamaroides          75
candidum               71
aligerum               69
bulbocastanum          67
pseudocapsicum         63
angustifolium          62
lycopersicum           60
umbelliferum           54
stenophyllidium        54
appendiculatum         49
hirtum                 49
mitlense               47
triquetrum             46
deflexum    

This is how many valid records we have for the species, but we still need to match them to valid multimedia values. Let's further reduce the data frame to just having the species so as to simplify it.

In [80]:
idigbio_ocurrence_species = idigbio_ocurrence_identified[["coreid", "dwc:specificEpithet"]]

idigbio_ocurrence_species.rename({
    "dwc:specificEpithet": "species"
}, axis=1, inplace=True)

idigbio_ocurrence_species.head(5)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


coreid        species
0  9dfd5d94-7f54-4e9f-880e-e683eae3a91c      houstonii
1  65c3037a-ea4b-4baa-81eb-8981c9a18b62  bulbocastanum
2  19f0db9c-e284-441a-b6ed-8e5a693ba506          grayi
3  0d53c48d-e5ad-43ba-8465-271b34cbdda9     nigrescens
4  a64aca3f-d22d-497e-9f18-74d2a2ade357     verrucosum

In [81]:
idigbio_ocurrence_species.shape

(6061, 2)

## Multimedia

Let's get the count of null values, taking special note of how the identifier column looks, this is the column where the URLs to the images are stored.

In [82]:
idigbio_multimedia.isna().sum()

coreid                         0
ac:accessURI                   0
idigbio:dateModified           0
idigbio:dataQualityScore       0
idigbio:etag                   0
idigbio:flags                  0
dcterms:format                20
idigbio:hasSpecimen            0
ac:licenseLogoURL            860
idigbio:mediaType           3109
dcterms:modified            7182
idigbio:recordIds              0
idigbio:records                0
idigbio:recordsets             0
dcterms:rights               860
ac:tag                      7051
dc:type                     1547
idigbio:uuid                   0
idigbio:version             7190
xmpRights:WebStatement       860
exif:PixelXDimension        4533
exif:PixelYDimension        4533
dtype: int64

Let's merge the species dataframe with the multimedia dataframe on the coreid

In [84]:
idigbio_images = idigbio_ocurrence_species.merge(idigbio_multimedia[["coreid", "ac:accessURI"]],
                                                 on="coreid",
                                                 how="inner")

display(idigbio_images.head(10))
print(idigbio_images.shape)

coreid  ...                                                                           ac:accessURI
0  9dfd5d94-7f54-4e9f-880e-e683eae3a91c  ...  http://herbarium.bio.fsu.edu/showimage.php?Image=images/herbarium/jpegs/000065632.jpg
1  65c3037a-ea4b-4baa-81eb-8981c9a18b62  ...                 http://data.huh.harvard.edu/9e9bfe45-e1a3-4ce0-8c2b-b28486b490b3/image
2  19f0db9c-e284-441a-b6ed-8e5a693ba506  ...                       http://n2t.net/ark:/65665/m3ae3960ec-d886-4eb8-b665-bb50c64c66f0
3  19f0db9c-e284-441a-b6ed-8e5a693ba506  ...                       http://n2t.net/ark:/65665/m3c1f6d537-d218-4012-8537-fd96fd180649
4  0d53c48d-e5ad-43ba-8465-271b34cbdda9  ...                       http://n2t.net/ark:/65665/m3cebed29a-30d2-43b9-8144-209b8c435a25
5  a64aca3f-d22d-497e-9f18-74d2a2ade357  ...                    http://ww2.bgbm.org/herbarium/images/B/10/01/88/09/B_10_0188092.jpg
6  469068c9-b52a-4dce-af43-37245149334e  ...                       http://n2t.net/ark:/65665/m3aadbad88-0089-48ac-af64-13261eea7762
7  469068c9-b52a-4dce-af43-37245149334e  ...                       http://n2t.net/ark:/65665/m36d1add6d-fe1f-48b8-9f83-b855be03cc40
8  67039f87-ec58-4699-9d67-dd9913975753  ...                       http://n2t.net/ark:/65665/m3a9c2f3b4-8f7c-48df-904f-5634ed179243
9  5489bccc-52a3-43c6-b31c-5fbc6e35ca9e  ...                       http://n2t.net/ark:/65665/m3bd4d23b0-704d-47e5-8d4a-bdb0102d00e2

[10 rows x 3 columns]

(6476, 3)


We need to check:

1. Why are we getting more records than on the dataframe for species? This is likely related to a given ID having more than one picture.
1. Check if there are null URLs, if there are we must remove them.

### Null URL's

In [85]:
idigbio_images.isna().sum()

coreid          0
species         0
ac:accessURI    0
dtype: int64

There are apparently no null reccords for the url.

### Check for multiple images

As stated before, we found that after the merge we had more records with multimedia than with species, let's figure out why.

This is most likely related to the fact that there might be multiple images for the same gbifID. This is fine and actually desirable.

In [86]:
idigbio_images["coreid"].value_counts()

722c3701-53e0-4a22-b40f-9f92780ef5b1    16
f3b64a1a-c9be-432d-a181-8b9f629efb94    12
b336a5f3-a3ed-4413-b4ad-476b7c2a89d2    10
b3fb1e6e-cefe-4c87-b549-d8020e246494    10
2cc705b8-05e9-4949-868f-c48d42043c67    10
                                        ..
abfc6a0f-404f-41f1-a63c-0187d9d1a892     1
c298e05f-cb18-4480-927d-2d8f18c988d1     1
fd02ac04-3a2c-4bc2-b79d-0cede018db45     1
31dad01b-ab93-4b5d-bb5c-097d5d557d32     1
01616322-7afb-4957-a209-4ba853b5770e     1
Name: coreid, Length: 6061, dtype: int64

### Duplicated URLs?

In [87]:
print(idigbio_images.shape)
print(len(idigbio_images["ac:accessURI"].unique()))

(6476, 3)
6476


In [88]:
idigbio_images[idigbio_images["ac:accessURI"].duplicated()]

Empty DataFrame
Columns: [coreid, species, ac:accessURI]
Index: []

We appear to have unique URLs. So we should be OK.

## Final count

We have a multimedia data frame, but after removing invalid species, removing records with no pictures, we must now figure out how many samples of each species we have and which species in Mexico we don't have a picture for.

In [89]:
display(idigbio_images["species"].value_counts())
print(idigbio_images.shape)

lanceolatum           371
stoloniferum          267
erianthum             256
americanum            245
pubigerum             242
nigrescens            225
ferrugineum           207
elaeagnifolium        206
houstonii             201
nigricans             188
nudum                 164
umbellatum            164
rostratum             163
hindsianum            159
diphyllum             135
chrysotrichum         130
torvum                126
demissum              118
rudepannum            111
aphyodendron          106
verrucosum            102
cardiophyllum          99
grayi                  83
refractum              83
candidum               81
schlechtendalianum     78
dulcamaroides          75
aligerum               71
pseudocapsicum         69
stenophyllidium        68
bulbocastanum          67
angustifolium          63
glaucescens            62
lycopersicum           61
umbelliferum           55
hirtum                 52
triquetrum             51
appendiculatum         50
citrullifoli

(6476, 3)


### We are missing these mexican species

In [90]:
missing_photo_samples = sorted([ms for ms in mexican_species if ms not in idigbio_images["species"].unique()])

print("\n".join(missing_photo_samples))
print(len(missing_photo_samples))

atitlanum
aviculare
bicorne
caripense
davisense
edmundoi
guerreroense
knoblochii
nitidibaccatum
setigeroides
triunfense
11


## Adding the section to the dataframe

In [109]:
sections = pd.read_csv("/content/drive/MyDrive/Tesis/Bibliografía/Secciones_Solanum.csv",
                       encoding="ISO-8859-1")

sections["Section"] = sections["Section"].apply(lambda sec: sec.split()[1].lower())

sections["Species"] = sections["Species"].apply(lambda sp: sp.split()[1].lower())

def fusion_species(series: pd.Series):
    unique_species = sorted(set(series))
    return ", ".join(unique_species)

species_by_section = sections.groupby('Section').agg({'Species':fusion_species}).reset_index()
display(species_by_section)

Section                                                                                                                                                                                                                                                                                                                                Species
0      acanthophora                                                                                                                                                                                                                                                                       acerifolium, aculeatissimum, capsicoides, mammosum, myriacanthum
1   anarrhichomenum                                                                                                                                                                                                                                                                            appendiculatum, ionidium, skutchii, subvelutinum, tacanense
2     archaesolanum                                                                                                                                                                                                                                                                                                                              aviculare
3        basarthrum                                                                                                                                                                                                                                                                                                       caripense, muricatum, suaveolens
4      brevantherum                                                                                                                                                                       atitlanum, axillifolium, brevipedicellatum, chiapasense, cordovense, erianthum, hazenii, lepidotum, plumense, pulverulentifolium, schlechtendalianum, umbellatum
5          crinitum                                                                                                                                                                                                                                                                                                                     mitlense, wrightii
6         dulcamara                                                                                                                                                                                                                                                      corymbosum, dulcamaroides, laxum, seaforthianum, sousae, triquetrum, umbelliferum
7    erythrotrichum                                                                                                                                                                                                                                                                                                                         erythrotrichum
8     gonatotrichum                                                                                                                                                                                                                                                                                                                   deflexum, lignescens
9      herposolanum                                                                                                                                                                                                                                                           bicorne, cobanense, edmundoi, glaucescens, refractum, triunfense, wendlandii
10    herpystrichum                                                                                                                                                                                

In [110]:
def determine_section(row: pd.Series, species_by_section: pd.DataFrame) -> str:

    specimen_species = row["species"]
    
    return sections[sections["Species"] == specimen_species]["Section"].to_string(index=False).strip()


idigbio_images["section"] = idigbio_images.apply(lambda row: determine_section(row, species_by_section), axis=1)

In [111]:
display(idigbio_images.isna().sum())
display(idigbio_images.head(20))

coreid          0
species         0
ac:accessURI    0
section         0
dtype: int64

coreid  ...        section
0   9dfd5d94-7f54-4e9f-880e-e683eae3a91c  ...      melongena
1   65c3037a-ea4b-4baa-81eb-8981c9a18b62  ...         petota
2   19f0db9c-e284-441a-b6ed-8e5a693ba506  ...      melongena
3   19f0db9c-e284-441a-b6ed-8e5a693ba506  ...      melongena
4   0d53c48d-e5ad-43ba-8465-271b34cbdda9  ...        solanum
5   a64aca3f-d22d-497e-9f18-74d2a2ade357  ...         petota
6   469068c9-b52a-4dce-af43-37245149334e  ...     holophylla
7   469068c9-b52a-4dce-af43-37245149334e  ...     holophylla
8   67039f87-ec58-4699-9d67-dd9913975753  ...          torva
9   5489bccc-52a3-43c6-b31c-5fbc6e35ca9e  ...  gonatotrichum
10  79053773-81cb-41ae-aea3-774d3340bcda  ...     holophylla
11  dfbda468-3c5e-41b1-9749-0079cbf6294f  ...     holophylla
12  329966cd-076a-44b6-8770-c2639d8d414f  ...         petota
13  5fd9f399-3316-4f99-88a5-e275e6fe4006  ...          torva
14  6787c1cb-5785-48e5-9027-cb7f1d831d10  ...         petota
15  0c5edbb1-22f6-4da1-bce4-addeee3609c1  ...   brevantherum
16  11ed5dcd-f482-4621-8087-0caaca38d645  ...         petota
17  269aa83f-faa4-4405-829d-8cc568e947cf  ...         petota
18  e6ed36f0-2e1d-4cc9-b3e3-42e20d68223b  ...          torva
19  d17d46bd-bdd3-4190-972a-8348f4c02329  ...          torva

[20 rows x 4 columns]

## How many samples per section do we have?

In [114]:
display(idigbio_images["section"].value_counts())

melongena          1129
petota             1083
holophylla         1024
torva              1008
brevantherum        651
solanum             474
dulcamara           260
herposolanum        179
lasiocarpa          142
micracantha         112
anarrhichomenum      72
acanthophora         64
lycopersicon         61
gonatotrichum        60
pachyphylla          58
crinitum             53
basarthrum           24
periscariae          14
erythrotrichum        3
pteroidea             3
herpystrichum         2
Name: section, dtype: int64

## What sections are we missing?

In [115]:
print("\n".join([s for s in sections["Section"].unique() if s not in idigbio_images["section"].unique()]))

archaesolanum


This is a really rare section, so it sounds reasonable that there are no speciments.

With this we can see that even though we don't have a sample of each specimen, we do have at least one specimen of each section.

## Store results to disk

In [116]:
idigbio_images.head(5)

coreid  ...    section
0  9dfd5d94-7f54-4e9f-880e-e683eae3a91c  ...  melongena
1  65c3037a-ea4b-4baa-81eb-8981c9a18b62  ...     petota
2  19f0db9c-e284-441a-b6ed-8e5a693ba506  ...  melongena
3  19f0db9c-e284-441a-b6ed-8e5a693ba506  ...  melongena
4  0d53c48d-e5ad-43ba-8465-271b34cbdda9  ...    solanum

[5 rows x 4 columns]

In [119]:
# Rename the identifier column
idigbio_images = idigbio_images.rename(columns={
    'coreid': 'id',
    'ac:accessURI': 'url'
    })

idigbio_images["source"] = "idigbio"

idigbio_images.head(3)

id        species  ...    section   source
0  9dfd5d94-7f54-4e9f-880e-e683eae3a91c      houstonii  ...  melongena  idigbio
1  65c3037a-ea4b-4baa-81eb-8981c9a18b62  bulbocastanum  ...     petota  idigbio
2  19f0db9c-e284-441a-b6ed-8e5a693ba506          grayi  ...  melongena  idigbio

[3 rows x 5 columns]

In [120]:
idigbio_images.to_csv(f"{idigbio_dataset_root_path}/idigbio_images_by_sections.csv",
                      index=False)

# Downloading of the images TBD

For this part, we need to use the data frame of images that we just built and go image by image attempting to download it. Images are linked by their URLs in two different ways:

1. As a direct link to the image file
1. As an HTML with a the image file linked in the `src` property of its `<img>` tag.

We should detect which type has been served to us so that we can act appropriately.

Images should be stored in a `<section>/` folder and have its file name following this template: `idigbio_<id>_<section>_<species>.jpg|png|gif`.

If a file couldn't be successfully pulled, the record should be put as it is in a new Data Frame for files markes as problematic and its final shape should be output to a new csv file called `idigbio_failed_records.csv`.

Prior to downloading, we should check if the file we intend to download has already been downloaded or not. If a file already exists, we should check if there are more files to download for the same specimen, if there are, an index should be suffixed at the end of the filename, if not, no index should be added.

In [ ]:
idigbio_images.read_csv(f"{idigbio_dataset_root_path}/idigbio_images_by_sections.csv")
output_path = ""

In [ ]:
for index, row in df.iterrows():
    print(row['c1'], row['c2'])